apps/extension/src/background/approvals/service.test.ts

```diff
@@ -0,0 +1,481 @@
+  let keyRingService: jest.Mocked<KeyRingService>;
+  let dataStore: KVStoreMock<string>;
+  let txStore: KVStoreMock<TxStore>;
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    txStore = new KVStoreMock<TxStore>("TxStore");
+    dataStore = new KVStoreMock<string>("DataStore");
+    const connectedTabsStore = new KVStoreMock<TabStore[]>("TabStore");
+    const approvedOriginsStore = new KVStoreMock<ApprovedOriginsStore>(
+      "ApprovedOriginsStore"
+    );
+    keyRingService = createMockInstance(KeyRingService as any);
+    const ledgerService: jest.Mocked<LedgerService> = createMockInstance(
+      LedgerService as any
+    );
+    const vaultService: jest.Mocked<VaultService> = createMockInstance(
+      VaultService as any
+    );
+
+    service = new ApprovalsService(
+      txStore,
+      dataStore,
+      connectedTabsStore,
+      approvedOriginsStore,
+      keyRingService,
+      ledgerService,
+      vaultService
+    );
+  });
+
+  describe("approveSignature", () => {
+    it("should add popupTabId to resolverMap", async () => {
+      const tabId = 1;
+      const sigResponse = {
+        hash: "hash",
+        signature: "sig",
+      };
+
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(tabId);
+      const signaturePromise = service.approveSignature("signer", "data");
+
+      await new Promise((resolve) =>
+        setTimeout(() => {
+          resolve(true);
+        })
+      );
+
+      (service as any).resolverMap[tabId].resolve(sigResponse);
+      const signature = await signaturePromise;
+
+      expect(signature).toEqual(sigResponse);
+    });
+
+    it("should throw an error when popupTabId is not present", async () => {
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(undefined);
+
+      expect(service.approveSignature("signer", "data")).rejects.toBeDefined();
+    });
+
+    it("should throw an error when popupTabId is already in the map", async () => {
+      const tabId = 1;
+      const sigResponse = {
+        hash: "hash",
+        signature: "sig",
+      };
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(tabId);
+      (service as any).resolverMap[tabId] = sigResponse;
+
+      expect(service.approveSignature("signer", "data")).rejects.toBeDefined();
+    });
+  });
+
+  describe("submitSignature", () => {
+    it("should add popupTabId to resolverMap", async () => {
+      const tabId = 1;
+      const sigResponse = {
+        hash: "hash",
+        signature: "sig",
+      };
+
+      jest.spyOn(service as any, "getPopupTabId").mockResolvedValue(tabId);
+      jest.spyOn(dataStore, "get").mockResolvedValueOnce("data");
+      jest
+        .spyOn(keyRingService, "signArbitrary")
+        .mockResolvedValue(sigResponse);
+      const signer = "signer";
+      const signaturePromise = service.approveSignature(signer, "data");
+
+      await new Promise((resolve) =>
```

Comment:
What does waiting here do?

Reply:
So the reason is `approveSignature` is async but it does not resolve the promise by itself.
Because of that I did not add `await` in front of the call as that would make test hang.
Instead we have to wait a "tick" so `resolverMap` will be filled and `submitSignature` will be able to resolve successfully 😅 It's a bit confusing.